refactor observables in core, cleanup and adding some samples

NikkyAI opened the pull request:
multiple observable maps were replaced with a single observable map and classes with observable values

fixes #1

ahinchman1 commented on core/src/main/kotlin/knote/api/NotebookManager.kt:
```diff
@@ -0,0 +1,37 @@
+                .mapValues { it.value.reports }
+                .filterValues { it != null }
+                .mapValues { it.value!! }
+        }
```
this is not a suggestion or anything, but seeing that these items are really similar in implementation, I wonder if this could be a point for refactoring later on but I'm not sure because it makes sense to keep the reporting separate

ahinchman1 replied:
Always better to have duplicate code for clarity instead of overdrying though, let's leave it alone.

NikkyAI replied:
mapValues and filterNotNull or a merged version would certainly be helpful

ahinchman1 commented on core/src/main/kotlin/knote/data/NotebookImpl.kt:
```diff
@@ -0,0 +1,32 @@
+    override val fileObject: MutableKObservableObject<Notebook, File> = MutableKObservableObject(file)
+    override val fileContentObject: MutableKObservableObject<Notebook, String> = MutableKObservableObject(fileContent)
+    override val compiledScriptObject: MutableKObservableObject<Notebook, NotebookScript?> =
+        MutableKObservableObject(compiledScript)
```
You can reduce some of the code here by going like 
```
override val fileObject = MutableKObservableObject<Notebook, File>(file) 
```
And so on

NikkyAI replied:
actually that does not work.. it needs to be declared as a MutableKObservable, because the interface just specifies a KObservable,
if i leave out the type here i have to cast this every time i want to mutate from within this object

ahinchman1 replied:
ahhhh okay. good catch! never mind on that

ahinchman1 commented on core/src/main/kotlin/knote/data/NotebookImpl.kt:
```diff
@@ -0,0 +1,32 @@
+    override var fileContent by fileContentObject
+    override var compiledScript by compiledScriptObject
+    override var reports by reportsObject
+    override var pageManager by pageManagerObject
```
Hmmm I really like the delegation!

NikkyAI replied:
i just wish this worked on the interface too

ahinchman1 commented on core/src/main/kotlin/knote/data/PageImpl.kt:
```diff
@@ -0,0 +1,33 @@
+    reports: List<ScriptDiagnostic>? = null
+) : Page {
+    override val fileObject: MutableKObservableObject<Page, File> = MutableKObservableObject(file)
+    override val fileContentObject: MutableKObservableObject<Page, String> = MutableKObservableObject(fileContent)
```
Same suggestion here

ahinchman1 commented on core/src/main/kotlin/knote/script/NotebookConfiguration.kt:
```diff
@@ -64,47 +73,52 @@ class NotebookConfiguration : ScriptCompilationConfiguration({
+                    reports += ScriptDiagnostic("importAnnotations: $importAnnotations", ScriptDiagnostic.Severity.DEBUG)
+
+                    val sources = importAnnotations.map {
+                        rootDir.resolve("include").resolve(it.source)
```
Could we do something like
```
rootDir.resolve("include/${it.source}")
```
?

NikkyAI replied:
i like a resolve call for each file level, no need for string interpolation either
and i see there is still importAnnotations there.. thoose have to be wiped since they break dependencies

ahinchman1 commented on core/src/main/kotlin/knote/PageManagerImpl.kt:
```diff
@@ -0,0 +1,314 @@
+
+    override val pages: MutableKObservableMap<String, Page> = MutableKObservableMap()
+    override val compiledPages: MutableKObservableMap<String, PageScript> = MutableKObservableMap()
+    override val results: MutableKObservableMap<String, Any> = MutableKObservableMap()
```
Same bit here too i.e.

```
override val pages = MutableKObservableMap<String, Page>()
override val compiledPages = MutableKObservableMap<String, PageScript>()
override val results = MutableKObservableMap<String, Any>()
```

NikkyAI replied:
same as above.. if this was not a override val it would work.. but the type of the overriden value takes precedence it seems

ahinchman1 commented on core/src/main/kotlin/knote/PageManagerImpl.kt:
```diff
@@ -0,0 +1,314 @@
+            evalPage(it)
+        }
+        while (pages.any { (id, page) -> page.result == null && !(page as PageImpl).errored}) {
+            pages.filter { it.value.result == null }
```
I think we can do
```
pages.filterNotNull(it.value.result)
```
here

NikkyAI replied:
rather `filterNull`, this filters for pages that do not have a result set

ahinchman1 commented on core/src/main/kotlin/knote/PageManagerImpl.kt:
```diff
@@ -0,0 +1,314 @@
+        notebookScript.pageFiles.forEach {
+            evalPage(it)
+        }
+        while (pages.any { (id, page) -> page.result == null && !(page as PageImpl).errored}) {
```
What does `page.result == null` do here?

NikkyAI replied:
`result == null` is just for filtering pages that do not have a result yet.. so they need to execute or re-execute
it also checks if it errored otherwise it will force into a infinite loop when some pages do not compile

it seems hte forEach call there used to be a `assosciate` call

ahinchman1 commented on core/src/main/kotlin/knote/PageManagerImpl.kt:
```diff
@@ -0,0 +1,314 @@
+    private fun invalidateResult(id: String) {
+        val page = pages[id] as PageImpl
+        logger.debug("invalidating result for '$id'")
+        logger.debug("dependencies: ${page.dependencies}")
```
Not a big deal here but for ones that are really long you can do something like 

```
logger.debug("""
		invalidating result for '$id'	
		dependencies: ${page.dependencies}
	""".trimIndent())
```

NikkyAI replied:
except then you get the problem with multiline output in the log that looks ugly

ahinchman1 replied:
oh bummer. No problem!

ahinchman1 commented on core/src/main/kotlin/knote/PageManagerImpl.kt:
```diff
@@ -0,0 +1,314 @@
+//                require(parameterProcessFunc.returnType.isSubtypeOf(property.returnType)) {
+//                    logger.error("${parameterProcessFunc.returnType} is not a subtype of requested ${property.returnType}")
+//                    logger.error("on page: $pageId")
+//                    logger.info("parameter script: ${parameterScript.id}")
```
Are these larger chunks taken out totally or moved elsewhere? I may have missed it on the first look

NikkyAI replied:
this is just taken out.. until i might re-add it or not
ddependency on other pages used to be done with those arguments, but it was a bit much typing
so now you can just do `val sleepData: DataFrame by inject()`
i hope to make this even sleeker without a need to declaring the type all over again

NikkyAI replied:
its a bit like a DI system

NikkyAI committed:
make Notebooks easier observable

use observables in core
add samples
restructure how dependencies are specified for a page
